atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -120,6 +125,7 @@ public Void apply(Client client) {
     private void cas(Client client, Long oldVal, long newVal) {
```

Comment:
Should probably assert as many preconditions as we can and fatally exit if violated, dumping as much info as possible, thinking something along the lines of:

```

    private synchronized Error exit(String message) {
        MultipleRunningTimestampServiceError error = new MultipleRunningTimestampServiceError(message);
        log.error("Terminating AtlasDB client process due to fatal error: {}", message, error);
        log.error("Current timestamp state:\n"
                        + "  currentLimit: {}\n"
                        + "  clientPool: {}",
                this.currentLimit, this.clientPool, this.lastWriteException
        );
        log.error(ThreadDumps.programmaticThreadDump());
        error.printStackTrace();
        System.exit(-1);
        throw error;
    }

    private void cas(Client client, Long oldVal, long newVal) {
        if (!Thread.holdsLock(this)) {
            throw exit(String.format("Attempted to cas while not holding lock, "
                            + "expected old value %s, new value: %d", oldVal, newVal);
        }
```